File: itx-rs/itx-impl/src/repo/postgres/post.rs
use std::collections::HashMap;

use async_trait::async_trait;
use itx_contract::repo::error::RepoError;
use itx_contract::repo::post::{
    AuthorId, ListPostsQuery, NewPost, Post, PostId, PostPatch, PostRepo,
};
use sqlx::{PgPool, Postgres, Transaction};

pub struct PostgresPostRepo {
    pool: PgPool,
}

impl PostgresPostRepo {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }
}

fn err<E: std::fmt::Display>(e: E) -> RepoError {
    RepoError::Unknown(e.to_string())
}

async fn upsert_tags(
    tx: &mut Transaction<'_, Postgres>,
    names: &[String],
) -> Result<Vec<i64>, RepoError> {
    let mut ids = Vec::with_capacity(names.len());
    for name in names {
        let id: i64 = sqlx::query_scalar(
            "INSERT INTO tags (name) VALUES ($1) \
             ON CONFLICT (name) DO UPDATE SET name = EXCLUDED.name \
             RETURNING id",
        )
        .bind(name)
        .fetch_one(&mut **tx)
        .await
        .map_err(err)?;
        ids.push(id);
    }
    Ok(ids)
}

async fn link_post_tags(
    tx: &mut Transaction<'_, Postgres>,
    post_id: PostId,
    tag_ids: &[i64],
) -> Result<(), RepoError> {
    for tid in tag_ids {
        sqlx::query("INSERT INTO post_tags (post_id, tag_id) VALUES ($1, $2) ON CONFLICT DO NOTHING")
            .bind(post_id)
            .bind(tid)
            .execute(&mut **tx)
            .await
            .map_err(err)?;
    }
    Ok(())
}

async fn fetch_tags_for(
    pool: &PgPool,
    post_ids: &[PostId],
) -> Result<HashMap<PostId, Vec<String>>, RepoError> {
    if post_ids.is_empty() {
        return Ok(HashMap::new());
    }
    let rows: Vec<(PostId, String)> = sqlx::query_as(
        "SELECT pt.post_id, t.name \
         FROM post_tags pt JOIN tags t ON pt.tag_id = t.id \
         WHERE pt.post_id = ANY($1) \
         ORDER BY t.name",
    )
    .bind(post_ids)
    .fetch_all(pool)
    .await
    .map_err(err)?;

    let mut map: HashMap<PostId, Vec<String>> = HashMap::new();
    for (pid, name) in rows {
        map.entry(pid).or_default().push(name);
    }
    Ok(map)
}

#[async_trait]
impl PostRepo for PostgresPostRepo {
    async fn list(&self, query: ListPostsQuery) -> Result<Vec<Post>, RepoError> {
        let limit = if query.limit == 0 { 50 } else { query.limit as i64 };
        let offset = query.offset as i64;

        let rows: Vec<(PostId, AuthorId, String, String, time::OffsetDateTime)> =
            match query.author_id {
                Some(author_id) => sqlx::query_as(
                    "SELECT id, author_id, title, body, created_at \
                     FROM posts WHERE author_id = $1 \
                     ORDER BY id DESC LIMIT $2 OFFSET $3",
                )
                .bind(author_id)
                .bind(limit)
                .bind(offset)
                .fetch_all(&self.pool)
                .await
                .map_err(err)?,
                None => sqlx::query_as(
                    "SELECT id, author_id, title, body, created_at \
                     FROM posts ORDER BY id DESC LIMIT $1 OFFSET $2",
                )
                .bind(limit)
                .bind(offset)
                .fetch_all(&self.pool)
                .await
                .map_err(err)?,
            };

        let ids: Vec<PostId> = rows.iter().map(|r| r.0).collect();
        let mut tag_map = fetch_tags_for(&self.pool, &ids).await?;

        Ok(rows
            .into_iter()
            .map(|(id, author_id, title, body, created_at)| Post {
                id,
                author_id,
                title,
                body,
                tags: tag_map.remove(&id).unwrap_or_default(),
                created_at,
            })
            .collect())
    }

    async fn get(&self, id: PostId) -> Result<Option<Post>, RepoError> {
        let row: Option<(PostId, AuthorId, String, String, time::OffsetDateTime)> =
            sqlx::query_as(
                "SELECT id, author_id, title, body, created_at FROM posts WHERE id = $1",
            )
            .bind(id)
            .fetch_optional(&self.pool)
            .await
            .map_err(err)?;

        let Some((id, author_id, title, body, created_at)) = row else {
            return Ok(None);
        };
        let mut tag_map = fetch_tags_for(&self.pool, &[id]).await?;
        Ok(Some(Post {
            id,
            author_id,
            title,
            body,
            tags: tag_map.remove(&id).unwrap_or_default(),
            created_at,
        }))
    }

    async fn create(&self, input: NewPost) -> Result<Post, RepoError> {
        let mut tx = self.pool.begin().await.map_err(err)?;

        let (id, created_at): (PostId, time::OffsetDateTime) = sqlx::query_as(
            "INSERT INTO posts (author_id, title, body) VALUES ($1, $2, $3) \
             RETURNING id, created_at",
        )
        .bind(input.author_id)
        .bind(&input.title)
        .bind(&input.body)
        .fetch_one(&mut *tx)
        .await
        .map_err(err)?;

        let tag_ids = upsert_tags(&mut tx, &input.tags).await?;
        link_post_tags(&mut tx, id, &tag_ids).await?;

        tx.commit().await.map_err(err)?;

        Ok(Post {
            id,
            author_id: input.author_id,
            title: input.title,
            body: input.body,
            tags: input.tags,
            created_at,
        })
    }

    async fn update(
        &self,
        id: PostId,
        author_id: AuthorId,
        patch: PostPatch,
    ) -> Result<Option<Post>, RepoError> {
        let mut tx = self.pool.begin().await.map_err(err)?;

        let existing: Option<(PostId, AuthorId, String, String, time::OffsetDateTime)> =
            sqlx::query_as(
                "SELECT id, author_id, title, body, created_at \
                 FROM posts WHERE id = $1 AND author_id = $2 FOR UPDATE",
            )
            .bind(id)
            .bind(author_id)
            .fetch_optional(&mut *tx)
            .await
            .map_err(err)?;

        let Some((_, _, mut title, mut body, created_at)) = existing else {
            return Ok(None);
        };

        if let Some(t) = patch.title {
            title = t;
        }
        if let Some(b) = patch.body {
            body = b;
        }

        sqlx::query("UPDATE posts SET title = $1, body = $2 WHERE id = $3")
            .bind(&title)
            .bind(&body)
            .bind(id)
            .execute(&mut *tx)
            .await
            .map_err(err)?;

        let tags = if let Some(new_tags) = patch.tags {
            sqlx::query("DELETE FROM post_tags WHERE post_id = $1")
                .bind(id)
                .execute(&mut *tx)
                .await
                .map_err(err)?;
            let tag_ids = upsert_tags(&mut tx, &new_tags).await?;
            link_post_tags(&mut tx, id, &tag_ids).await?;
            new_tags
        } else {
            let rows: Vec<(String,)> = sqlx::query_as(
                "SELECT t.name FROM post_tags pt JOIN tags t ON pt.tag_id = t.id \
                 WHERE pt.post_id = $1 ORDER BY t.name",
            )
            .bind(id)
            .fetch_all(&mut *tx)
            .await
            .map_err(err)?;
            rows.into_iter().map(|r| r.0).collect()
        };

        tx.commit().await.map_err(err)?;

        Ok(Some(Post {
            id,
            author_id,
            title,
            body,
            tags,
            created_at,
        }))
    }

    async fn delete(&self, id: PostId, author_id: AuthorId) -> Result<bool, RepoError> {
        let result = sqlx::query("DELETE FROM posts WHERE id = $1 AND author_id = $2")
            .bind(id)
            .bind(author_id)
            .execute(&self.pool)
            .await
            .map_err(err)?;
        Ok(result.rows_affected() > 0)
    }
}

File: itx-go/itx-impl/repo/postgres/post.go
package postgres

import (
	"context"
	"database/sql"
	"errors"
	"strconv"
	"strings"
	"time"

	"github.com/chehsunliu/itx/itx-go/itx-contract/repo/post"
)

type postRepo struct {
	db *sql.DB
}

func (r *postRepo) List(ctx context.Context, params post.ListParams) ([]post.Post, error) {
	limit := int64(params.Limit)
	if limit == 0 {
		limit = 50
	}
	offset := int64(params.Offset)

	var (
		rows *sql.Rows
		err  error
	)
	if params.AuthorID != nil {
		rows, err = r.db.QueryContext(ctx,
			"SELECT id, author_id, title, body, created_at "+
				"FROM posts WHERE author_id = $1 "+
				"ORDER BY id DESC LIMIT $2 OFFSET $3",
			params.AuthorID.String(), limit, offset)
	} else {
		rows, err = r.db.QueryContext(ctx,
			"SELECT id, author_id, title, body, created_at "+
				"FROM posts ORDER BY id DESC LIMIT $1 OFFSET $2",
			limit, offset)
	}
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var posts []post.Post
	var ids []int64
	for rows.Next() {
		var p post.Post
		var authorID string
		if err := rows.Scan(&p.ID, &authorID, &p.Title, &p.Body, &p.CreatedAt); err != nil {
			return nil, err
		}
		if err := p.AuthorID.UnmarshalText([]byte(authorID)); err != nil {
			return nil, err
		}
		posts = append(posts, p)
		ids = append(ids, p.ID)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}

	tagMap, err := fetchTagsFor(ctx, r.db, ids)
	if err != nil {
		return nil, err
	}
	for i := range posts {
		posts[i].Tags = tagMap[posts[i].ID]
		if posts[i].Tags == nil {
			posts[i].Tags = []string{}
		}
	}
	return posts, nil
}

func (r *postRepo) Get(ctx context.Context, params post.GetParams) (post.Post, error) {
	row := r.db.QueryRowContext(ctx,
		"SELECT id, author_id, title, body, created_at FROM posts WHERE id = $1",
		params.ID)
	var p post.Post
	var authorID string
	if err := row.Scan(&p.ID, &authorID, &p.Title, &p.Body, &p.CreatedAt); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return post.Post{}, post.ErrNotFound
		}
		return post.Post{}, err
	}
	if err := p.AuthorID.UnmarshalText([]byte(authorID)); err != nil {
		return post.Post{}, err
	}

	tagMap, err := fetchTagsFor(ctx, r.db, []int64{p.ID})
	if err != nil {
		return post.Post{}, err
	}
	p.Tags = tagMap[p.ID]
	if p.Tags == nil {
		p.Tags = []string{}
	}
	return p, nil
}

func (r *postRepo) Create(ctx context.Context, params post.CreateParams) (post.Post, error) {
	tx, err := r.db.BeginTx(ctx, nil)
	if err != nil {
		return post.Post{}, err
	}
	defer tx.Rollback()

	var id int64
	var createdAt time.Time
	err = tx.QueryRowContext(ctx,
		"INSERT INTO posts (author_id, title, body) VALUES ($1, $2, $3) RETURNING id, created_at",
		params.AuthorID.String(), params.Title, params.Body).Scan(&id, &createdAt)
	if err != nil {
		return post.Post{}, err
	}

	tagIDs, err := upsertTagsTx(ctx, tx, params.Tags)
	if err != nil {
		return post.Post{}, err
	}
	if err := linkPostTagsTx(ctx, tx, id, tagIDs); err != nil {
		return post.Post{}, err
	}

	if err := tx.Commit(); err != nil {
		return post.Post{}, err
	}

	tags := append([]string{}, params.Tags...)
	return post.Post{
		ID:        id,
		AuthorID:  params.AuthorID,
		Title:     params.Title,
		Body:      params.Body,
		Tags:      tags,
		CreatedAt: createdAt,
	}, nil
}

func (r *postRepo) Update(ctx context.Context, params post.UpdateParams) (post.Post, error) {
	tx, err := r.db.BeginTx(ctx, nil)
	if err != nil {
		return post.Post{}, err
	}
	defer tx.Rollback()

	var p post.Post
	var authorID string
	err = tx.QueryRowContext(ctx,
		"SELECT id, author_id, title, body, created_at FROM posts "+
			"WHERE id = $1 AND author_id = $2 FOR UPDATE",
		params.ID, params.AuthorID.String()).Scan(&p.ID, &authorID, &p.Title, &p.Body, &p.CreatedAt)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return post.Post{}, post.ErrNotFound
		}
		return post.Post{}, err
	}
	p.AuthorID = params.AuthorID

	if params.Title != nil {
		p.Title = *params.Title
	}
	if params.Body != nil {
		p.Body = *params.Body
	}
	if _, err := tx.ExecContext(ctx,
		"UPDATE posts SET title = $1, body = $2 WHERE id = $3",
		p.Title, p.Body, p.ID); err != nil {
		return post.Post{}, err
	}

	if params.Tags != nil {
		if _, err := tx.ExecContext(ctx, "DELETE FROM post_tags WHERE post_id = $1", p.ID); err != nil {
			return post.Post{}, err
		}
		tagIDs, err := upsertTagsTx(ctx, tx, *params.Tags)
		if err != nil {
			return post.Post{}, err
		}
		if err := linkPostTagsTx(ctx, tx, p.ID, tagIDs); err != nil {
			return post.Post{}, err
		}
		p.Tags = append([]string{}, (*params.Tags)...)
	} else {
		rows, err := tx.QueryContext(ctx,
			"SELECT t.name FROM post_tags pt JOIN tags t ON pt.tag_id = t.id "+
				"WHERE pt.post_id = $1 ORDER BY t.name", p.ID)
		if err != nil {
			return post.Post{}, err
		}
		var tags []string
		for rows.Next() {
			var name string
			if err := rows.Scan(&name); err != nil {
				rows.Close()
				return post.Post{}, err
			}
			tags = append(tags, name)
		}
		rows.Close()
		if tags == nil {
			tags = []string{}
		}
		p.Tags = tags
	}

	if err := tx.Commit(); err != nil {
		return post.Post{}, err
	}
	return p, nil
}

func (r *postRepo) Delete(ctx context.Context, params post.DeleteParams) error {
	result, err := r.db.ExecContext(ctx,
		"DELETE FROM posts WHERE id = $1 AND author_id = $2",
		params.ID, params.AuthorID.String())
	if err != nil {
		return err
	}
	rows, err := result.RowsAffected()
	if err != nil {
		return err
	}
	if rows == 0 {
		return post.ErrNotFound
	}
	return nil
}

func upsertTagsTx(ctx context.Context, tx *sql.Tx, names []string) ([]int64, error) {
	ids := make([]int64, 0, len(names))
	for _, name := range names {
		var id int64
		if err := tx.QueryRowContext(ctx,
			"INSERT INTO tags (name) VALUES ($1) "+
				"ON CONFLICT (name) DO UPDATE SET name = EXCLUDED.name "+
				"RETURNING id", name).Scan(&id); err != nil {
			return nil, err
		}
		ids = append(ids, id)
	}
	return ids, nil
}

func linkPostTagsTx(ctx context.Context, tx *sql.Tx, postID int64, tagIDs []int64) error {
	for _, tagID := range tagIDs {
		if _, err := tx.ExecContext(ctx,
			"INSERT INTO post_tags (post_id, tag_id) VALUES ($1, $2) ON CONFLICT DO NOTHING",
			postID, tagID); err != nil {
			return err
		}
	}
	return nil
}

func fetchTagsFor(ctx context.Context, db *sql.DB, ids []int64) (map[int64][]string, error) {
	out := map[int64][]string{}
	if len(ids) == 0 {
		return out, nil
	}
	placeholders := make([]string, len(ids))
	args := make([]any, len(ids))
	for i, id := range ids {
		placeholders[i] = "$" + strconv.Itoa(i+1)
		args[i] = id
	}
	rows, err := db.QueryContext(ctx,
		"SELECT pt.post_id, t.name FROM post_tags pt JOIN tags t ON pt.tag_id = t.id "+
			"WHERE pt.post_id IN ("+strings.Join(placeholders, ",")+") ORDER BY t.name",
		args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	for rows.Next() {
		var pid int64
		var name string
		if err := rows.Scan(&pid, &name); err != nil {
			return nil, err
		}
		out[pid] = append(out[pid], name)
	}
	return out, rows.Err()
}
